### PR TITLE
(MISC): avoid unnecessary `subSequence` calls

### PR DESCRIPTION
**Pull request by jajakobyly**
See discussion under 9970c82fb72d28b37ca434356cf4868880aec900

### CODE REVIEW

**Comment by jajakobyly** [comment]
@matklad @e-oz please check if it doesn't accidentally recreate #367

**Comment by e-oz** [comment]
@jajakobyly last nightly build works fine

**Comment by matklad** [comment]
@jajakobyly I think it should be possible to reuse `indexOf` from `com.intellij.util.text.CharArrayUtil` here.

**Comment by matklad** [comment]
I mean, to reuse it *inside* the extension method.

**Comment by matklad** [comment]
TIL:

It is unspecified if `CharSequence::subSequence` is O(1) or O(N). At least for `String` since Java 7 it is O(N).

If you want an explicit O(1) sub sequnece, use `CharSequenceSubSequence`. 

`ImmutableText`, which is used in IDEA, has O(1) subSequence.

`CharArrayUtil` has some useful stuff!

**Comment by jajakobyly** [comment]
Ok pushed `CharArrayUtil`'s `indexOf` as an static import, because introducing special extension method looked like a triumph of form over content for me.

**Comment by jajakobyly** [comment]
@matklad looks like the `com.intellij.util.text` package has lots of useful stuff :)

File: src/main/kotlin/org/rust/lang/core/lexer/RustEscapesLexer.kt [hunk 1]
```diff
@@ -82,18 +84,11 @@ class RustEscapesLexer private constructor(val defaultToken: IElementType,
             }
             return i + 1
         } else {
-            val idx = indexOf('\\', start + 1)
-            return if (idx != -1) Math.min(idx, bufferEnd) else bufferEnd
+            val idx = indexOf(bufferSequence, "\\", start + 1, bufferEnd)
+            return if (idx != -1) min(idx, bufferEnd) else bufferEnd
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/lexer/RustEscapesLexer.kt]: I bet this min always returns `idx`.

File: src/main/kotlin/org/rust/lang/core/lexer/RustEscapesLexer.kt [hunk 2]
```diff
@@ -60,17 +62,17 @@ class RustEscapesLexer private constructor(val defaultToken: IElementType,
...
                         } else {
                             return i + 1 + 1
                         }
                     }
                 'u'        ->
                     if (bufferEnd - (i + 1) >= 1 && bufferSequence[i + 1] == '{') {
-                        val idx = indexOf('}', i + 1)
-                        return if (idx != -1) Math.min(idx + 1, bufferEnd) else bufferEnd
+                        val idx = indexOf(bufferSequence, "}", i + 1, bufferEnd)
+                        return if (idx != -1) min(idx + 1, bufferEnd) else bufferEnd
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/lexer/RustEscapesLexer.kt]: and also this one.

**Comment by matklad** [comment]
>looks like the com.intellij.util.text package has lots of useful stuff :)

Yep, will use `SevVer` for [this thing](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/cargo/project/CargoProjectDescription.kt#L22) one day

File: src/main/kotlin/org/rust/lang/core/lexer/RustEscapesLexer.kt (hunk 1 above)
Comment (jajakobyly) [inline src/main/kotlin/org/rust/lang/core/lexer/RustEscapesLexer.kt]: Yeah, you're right

**Comment by jajakobyly** [comment]
Done

**Comment by matklad** [comment]
Thanks!

**Pull request merged**